pmd-core/src/main/java/net/sourceforge/pmd/PmdAnalysis.java

```diff
@@ -416,10 +416,20 @@ private static GlobalAnalysisListener createComposedRendererListener(List<Render
         return GlobalAnalysisListener.tee(rendererListeners);
     }
 
-    private Set<Language> getApplicableLanguages(boolean quiet) {
+        private Set<Language> getApplicableLanguages(boolean quiet) {
         Set<Language> languages = new HashSet<>();
         LanguageVersionDiscoverer discoverer = configuration.getLanguageVersionDiscoverer();
 
+        // Extract nested loop into a separate method
+        collectApplicableLanguages(ruleSets, languages, discoverer, quiet);
```

Comment:
Two new methods are introduced to reduce the complexity of getApplicableLanguages() method.